### PR TITLE
[PyOV] Added deprecation function

### PR DESCRIPTION
**Pull request by adrianboguszewski**
### Details:
 - *Add deprecation warning in c++ common namespace*
 - *Add python binding*
 - *Add python deprecation decorator*

### Tickets:
 - *82660*

### CODE REVIEW

File: src/bindings/python/src/openvino/utils.py [hunk 1]
```diff
@@ -28,3 +29,27 @@ def add_openvino_libs_to_path() -> None:
...
+        @wraps(wrapped)
+        def wrapper(*args: Any, **kwargs: Any) -> Callable[..., Any]:
+            from openvino.pyopenvino.util import deprecation_warning
+            deprecation_warning(wrapped.__name__, version, message)
+            return wrapped(*args, **kwargs)
+        return wrapper
+
+    # decorated function is assigned to version if there are no arguments passed
+    arg1 = version
+    if callable(arg1):
```
Comment (p-wysocki) [inline src/bindings/python/src/openvino/utils.py:50]: Can we change `arg1` to something more descriptive? Also, type hint in L40 does not `version` could be `Callable`, just a string. I think I am missing something, why should it be checked if `version` is callable?

Did you mean that the first argument can be either a Python function object or a version, and branch `deprecated()` based on that? Do we expect `version` to be a function object?

File: src/bindings/python/src/openvino/utils.py [hunk 2]
```diff
@@ -28,3 +29,27 @@ def add_openvino_libs_to_path() -> None:
                 # On Windows, with Python >= 3.8, DLLs are no longer imported from the PATH.
                 if (3, 8) <= sys.version_info:
                     os.add_dll_directory(os.path.abspath(lib_path))
+
+
+def deprecated(version: str = "", message: str = "") -> Callable[..., Any]:
```
Comment (p-wysocki) [inline src/bindings/python/src/openvino/utils.py:34]: This decorator seems a bit heavy, wouldn't something like this be sufficient?
```suggestion
def deprecated(wrapped: Callable[..., Any], version: str = "", message: str = "") -> Callable[..., Any]:
    """Prints deprecation warning and runs the function.
    :param wrapped: The function to decorate
    :param version: The version in which the code will be removed.
    :param message: A message explaining why the function is deprecated and/or what to use instead.
    """
    @wraps(wrapped)
    def wrapper(*args: Any, **kwargs: Any):
        deprecation_warning(wrapped.__name__, version, message)
        return wrapped(*args, **kwargs)
    return wrapper
```

**Review by p-wysocki — CHANGES_REQUESTED** [review]

**Comment by p-wysocki** [comment]
We also could use a binding test, probably somewhere in `openvino/src/bindings/python/tests/test_utils/test_utils.py`.
Additionally, it would be great to use this function in places where we use old, inconsistent deprecation warnings - it has already been discussed with @adrianboguszewski and is work in progress.

File: src/bindings/python/src/openvino/utils.py [hunk 3]
```diff
@@ -1,9 +1,10 @@
 # -*- coding: utf-8 -*-
 # Copyright (C) 2021 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
-
```
[the remark this answers is not in the document]
Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py:4]: It was deleted automatically by IDE. I'll restore that.

File: src/bindings/python/src/openvino/utils.py [hunk 4]
```diff
@@ -28,3 +29,27 @@ def add_openvino_libs_to_path() -> None:
...
+def deprecated(version: str = "", message: str = "") -> Callable[..., Any]:
+    """Prints deprecation warning and runs the function.
+
+    :param version: The version in which the code will be removed.
+    :param message: A message explaining why the function is deprecated and/or what to use instead.
+    """
+    def decorator(wrapped: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(wrapped)
+        def wrapper(*args: Any, **kwargs: Any) -> Callable[..., Any]:
+            from openvino.pyopenvino.util import deprecation_warning
```
[the remark this answers is not in the document]
Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py:43]: It should. I was testing something and forgot to move it back

File: src/bindings/python/src/pyopenvino/utils/utils.hpp [hunk 5]
```diff
@@ -18,6 +18,8 @@ namespace utils {
     std::map<std::string, ov::Any> properties_to_any_map(const std::map<std::string, py::object>& properties);
 
     std::string convert_path_to_string(const py::object& path);
+
+    void deprecation_warning(const std::string& function_name, const std::string& version = std::string(), const std::string& message = std::string());
```
[the remark this answers is not in the document]
Comment (adrianboguszewski) [reply src/bindings/python/src/pyopenvino/utils/utils.hpp:22]: If not empty string as a default argument, then what?

File: src/bindings/python/src/openvino/utils.py (hunk 1 above)
Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py:50]: This is quite complicated. If you add the decorator with arguments (`@deprecated(message="msg", version="2.3")` ) version is a string, but if you add the decorator with no arguments (`@deprecated`) the version is the function it decorates

File: src/bindings/python/src/openvino/utils.py (hunk 2 above)
Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py:34]: Nope. It cannot be like that. It won't work if you add the 'wrapped' argument to the function up:
```Traceback (most recent call last):
  File "/home/adrian/PycharmProjects/pythonProject/main.py", line 20, in <module>
    @deprecated(message="msg")
TypeError: deprecated() missing 1 required positional argument: 'wrapped'```

Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py:34]: see Example #3 here: https://www.geeksforgeeks.org/decorators-with-parameters-in-python/

File: src/bindings/python/src/openvino/utils.py (hunk 1 above)
Comment (adrianboguszewski) [reply src/bindings/python/src/openvino/utils.py]: We decided to simplify the implementation

**Review by p-wysocki — APPROVED** [review]
LGTM

File: src/bindings/python/src/pyopenvino/graph/passes/manager.cpp [hunk 6]
```diff
@@ -69,10 +68,9 @@ void regclass_passes_Manager(py::module m) {
     manager.def(
         "register_pass",
         [](ov::pass::Manager& self, const std::string& pass_name) -> void {
-            PyErr_WarnEx(PyExc_DeprecationWarning,
-                         "register_pass with this arguments is deprecated! "
-                         "Please use register_pass(ConstantFolding()) instead.",
-                         1);
+            Common::utils::deprecation_warning("register_pass(pass_name)",
+                                               "",
+                                               "Please use register_pass(ConstantFolding()) instead.");
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/graph/passes/manager.cpp:73]: Can this be tested as well?

**Review by jiwaszki — CHANGES_REQUESTED** [review]

File: src/bindings/python/src/pyopenvino/graph/passes/manager.cpp (hunk 6 above)
Comment (adrianboguszewski) [reply src/bindings/python/src/pyopenvino/graph/passes/manager.cpp:73]: I cannot see any c++ tests (except mocks) in the python bindings directory. I think this function is indirectly tested by testing python binding and decorator. However, if you want a separate test, please point me to the right place for such a test.

**Review by jiwaszki — APPROVED** [review]

**Review by jiwaszki — APPROVED** [review]

**Pull request merged**